Add missing `layer(…)` to imports above Tailwind directives

RobinMalfait opened the pull request:
This PR fixes an issue where imports above Tailwind directives didn't get a `layer(…)` argument.

Given this CSS:
```css
@import "./typography.css";
@tailwind base;
@tailwind components;
@tailwind utilities;
```

It was migrated to:
```css
@import "./typography.css";
@import "tailwindcss";
```

But to ensure that the typography styles end up in the correct location, it requires the `layer(…)` argument.

This PR now migrates the input to:
```css
@import "./typography.css" layer(base);
@import "tailwindcss";
```

Test plan:
---

Added an integration test where an import receives the `layer(…)`, but an import that eventually contains `@utility` does not receive the `layer(…)` argument. This is necessary otherwise the `@utility` will be nested when we are processing the inlined CSS.

Running this on the Commit template, we do have a proper `layer(…)`
<img width="585" alt="image" src="https://github.com/user-attachments/assets/538055e6-a9ac-490d-981f-41065a6b59f9">

RobinMalfait committed:
add failing test

RobinMalfait committed:
add `layer(…)` to `@import` at-rules when necessary

RobinMalfait committed:
update changelog

RobinMalfait commented on integrations/upgrade/index.test.ts:
```diff
@@ -453,6 +453,130 @@ test(
+        color: red;
+      }
+
+      html {
+        color: blue;
+      }
```
Something to think about: should this file be split into an import _with_ the `layer(…)` and one with the utilities to ensure that this `html {}` base CSS is in the correct spot. 🤔

RobinMalfait replied:
I think this is right, otherwise `html {}` would be in the correct spot. I implemented it in the followup commits.

RobinMalfait committed:
use Stylesheets containing `@utility`

At this point it doesn't matter if we are layered or not, but if the
stylesheet contains an `@utility`. If it does, then that's a reason to
potentially split stylesheets (if necessary).

RobinMalfait commented on integrations/upgrade/index.test.ts:
```diff
@@ -1571,8 +1698,9 @@ test(
 
       --- ./src/components.css ---
-      @import './typography.css';
+      @import './typography.css' layer(components);
```
This has `layer(components)` because this is coming from the components file where nothing sits above it (in that file), but there is a `@layer components` and `@tailwind components` below it.

_if_ we want to look at as-if it was flattened, then it exists between `base` and `components`. This means that if this code existed in the same file we would've generated `layer(base)` instead. 🤔

RobinMalfait replied:
This is also what we get on the Commit template with this PR:
<img width="585" alt="image" src="https://github.com/user-attachments/assets/a4ae9f4d-44da-4afe-9d48-dc434bfc948d">

Without knowing the rest of the files, it looks correct. But I wonder if `layer(base)` is more correct.

RobinMalfait replied:
Talked with @philipp-spiess and @thecrypticace and we came to the conclusion that this is the least surprising outcome where you look at CSS files in isolation.

philipp-spiess reviewed (commented):
Maybe we can fix the case where the `tailwind-setup.css` file is split, otherwise this looks solid

RobinMalfait committed:
root files don't need to be split

When we detect CSS files that contain `@utility` and non-`@utility`
parts, then we split them. We will also split every parent file to
ensure the imports are correct.

The new files with `@utility` should not result in a `layer(…)` next to
the imports. The other CSS might end up with a `layer(…)`

RobinMalfait committed:
update integration tests

RobinMalfait commented on integrations/upgrade/index.test.ts:
```diff
@@ -1706,16 +1836,19 @@ test(
       }
 
       --- ./src/components.css ---
-      @import './typography.css';
+      @import './typography.css' layer(components);
 
+      --- ./src/components.utilities.css ---
       @utility foo {
         color: red;
       }
```
I think that technically since the only non `@utility` is an `@import` _with_ a `layer(…)` that this split shouldn't be required. But that feels like an optimization for a separate PR (if we even need that.).

RobinMalfait commented on integrations/upgrade/index.test.ts:
```diff
@@ -1584,6 +1712,10 @@ test(
 
+      --- ./src/tailwind-setup.utilities.css ---
+      @import './components.utilities.css';
+      @import './utilities.utilities.css'
```
Good catch, _a_ split made sense, but the split should've stopped at the Tailwind root file indeed. Handled here: https://github.com/tailwindlabs/tailwindcss/pull/14982/commits/b62c49990dfdc3aa37fb0b6b2afaf8b3e6d6dc41

You can see the difference in the integration tests: https://github.com/tailwindlabs/tailwindcss/pull/14982/commits/87c0bb29e55a939cad795ad318c6300d65c4452e

Re-ran it on the Commit template and the output is the same (because no splits happened there, but it also didn't break anything)

RobinMalfait committed:
improve sheet splitting

We only need to split if the stylesheet contains `@utility` and non-safe
CSS. Imports with `layer(…)`, `@layer` blocks and comments are
considered safe, everything else is not.

The non safe-code might end up in a new import with a `layer(…)` to
guarantee that it's injected in the correct spot.

RobinMalfait committed:
update integration tests with improved logic
